host_applications/linux/apps/raspicam/RaspiGPS.c

```diff
@@ -0,0 +1,188 @@
+#include <errno.h>
+#include <sysexits.h>
+
+#include "RaspiGPS.h"
+
+typedef struct
+{
+   pthread_mutex_t gps_cache_mutex;
+   gpsd_info gpsd;
+   struct gps_data_t gpsdata_cache;
+   time_t last_valid_time;
+   pthread_t gps_reader_thread;
+   int terminated;
+   int gps_reader_thread_ok;
+} GPS_READER_DATA;
+
+static GPS_READER_DATA gps_reader_data;
+
+#define GPS_CACHE_EXPIRY      5 // in seconds
+
+struct gps_data_t *raspi_gps_lock()
+{
+   pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+   return &gps_reader_data.gpsdata_cache;
+}
+
+void raspi_gps_unlock()
+{
+   pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+}
+
+
+static void *gps_reader_process(void *gps_reader_data_ptr)
+{
+   while (!gps_reader_data.terminated)
+   {
+      int ret = 0;
+      gps_reader_data.gpsd.gpsdata.set = 0;
+      gps_reader_data.gpsd.gpsdata.fix.mode = 0;
+      if ((connect_gpsd(&gps_reader_data.gpsd) < 0) ||
+            ((ret = read_gps_data_once(&gps_reader_data.gpsd)) < 0))
+         break;
+
+      int gps_valid = 0;
+      if ((ret > 0) && (gps_reader_data.gpsd.gpsdata.online))
+      {
+         if (gps_reader_data.gpsd.gpsdata.fix.mode >= MODE_2D)
+         {
+            // we have GPS fix, copy fresh data to cache
+            gps_valid = 1;
+            time(&gps_reader_data.last_valid_time);
+            pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+            memcpy(&gps_reader_data.gpsdata_cache, &gps_reader_data.gpsd.gpsdata,
+                   sizeof(struct gps_data_t));
+            pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+         }
+      }
+      if (!gps_valid)
+      {
+         time_t now;
+         time(&now);
+         if (now - gps_reader_data.last_valid_time > GPS_CACHE_EXPIRY)
+         {
+            // our cache is stale, clear it
+            pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+            gps_reader_data.gpsdata_cache.online = gps_reader_data.gpsd.gpsdata.online;
+            gps_reader_data.gpsdata_cache.set = 0;
+            gps_reader_data.gpsdata_cache.fix.mode = 0;
+            pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+         }
+         // we lost GPS fix, copy GPS time to cache if available
+         if (gps_reader_data.gpsd.gpsdata.set & TIME_SET)
+         {
+            pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+            gps_reader_data.gpsdata_cache.set |= TIME_SET;
+            gps_reader_data.gpsdata_cache.fix.time = gps_reader_data.gpsd.gpsdata.fix.time;
+            pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+         }
+      }
+   }
+   return NULL;
+}
+
+void raspi_gps_shutdown(int verbose)
+{
+   gps_reader_data.terminated = 1;
+
+   if (gps_reader_data.gps_reader_thread_ok)
+   {
+      if (verbose)
+         fprintf(stderr, "Waiting for GPS reader thread to terminate\n");
+
+      pthread_join(gps_reader_data.gps_reader_thread, NULL);
+   }
+   if (verbose && (gps_reader_data.gpsd.gpsd_connected))
```

Comment:
I see they've been copied from the original but the extra parentheses just make it look more complex than it is.